apluslms_roman/backends/docker.py

```diff
@@ -76,29 +76,32 @@ def _run_opts(self, task, step):
 
     def prepare(self, task, observer):
         client = self._client
-        for step in task.steps:
-            observer.step_preflight(step)
-            image, tag = step.img.split(':', 1)
-            try:
-                img = client.images.get(step.img)
-            except docker.errors.ImageNotFound:
-                observer.step_running(step)
-                observer.manager_msg(step, "Downloading image {}".format(step.img))
+        try:
+            for step in task.steps:
+                observer.step_preflight(step)
+                image, tag = step.img.split(':', 1)
                 try:
-                    img = client.images.pull(image, tag)
-                except docker.errors.APIError as err:
-                    observer.step_failed(step)
-                    error = "%s %s" % (err.__class__.__name__, err)
-                    return BuildResult(-1, error, step)
-            observer.step_succeeded(step)
-        return BuildResult()
+                    img = client.images.get(step.img)
+                except docker.errors.ImageNotFound:
+                    observer.step_running(step)
+                    observer.manager_msg(step, "Downloading image {}".format(step.img))
+                    try:
+                        img = client.images.pull(image, tag)
+                    except docker.errors.APIError as err:
+                        observer.step_failed(step)
+                        error = "%s %s" % (err.__class__.__name__, err)
+                        return BuildResult(step, False, -1, error)
+                observer.step_succeeded(step)
+        except KeyboardInterrupt:
+            return BuildResult(step, False)
```

Comment:
I personally don't like to refer variables defined in for statement, outside of the for block. So I would probably wrap inside of the for block.

In all cases, there are same times when SIGTERM can be received, outside of the try-except blocks, so I guess we need one in the builder to capture those. Of course, there we won't have step info, but it kind of means that we don't know which step was the last.

Hmm.. That actually means that my code sample for result_msg needs to work bit differently. It shouldn't update steps which state is `completed`.

Reply:
Should I just wrap most of Builder.build in a try-except block?

Reply:
You could try it out at least :)